eclass/dist-kernel-utils.eclass

```diff
@@ -66,18 +66,56 @@ dist-kernel_get_image_path() {
 }
 
 # @FUNCTION: dist-kernel_install_kernel
-# @USAGE: <version> <image> <system.map>
+# @USAGE: <version> <image> <system.map> <dir> [--optargs]
 # @DESCRIPTION:
-# Install kernel using installkernel tool.  <version> specifies
-# the kernel version, <image> full path to the image, <system.map>
-# full path to System.map.
+# Install kernel using installkernel tool. <version> specifies the
+# kernel version, <image> full path to the image, <system.map> full
+# path to System.map, and <dir> the install target directory. If
+# unspecified then <version> defaults to ${KV_FULL}, <image> defaults
+# to the standard kernel image at ${KV_DIR}, <system.map> defaults to
+# the System.map file in the ${KV_DIR}, and <dir> defaults to
+# ${EROOT}/boot. Any additional optional arguments are directly passed
+# on to installkernel if the installed version of installkernel is
+# at least 56. See the installkernel and kernel-install manuals for
+# an overview of which optional arguments are supported. Of particular
+# interest is the --all argument which may be used to iteratively
+# (re-)install all distribution kernels installed on the system.
 dist-kernel_install_kernel() {
 	debug-print-function ${FUNCNAME} "$@"
 
-	[[ ${#} -eq 3 ]] || die "${FUNCNAME}: invalid arguments"
-	local version=${1}
-	local image=${2}
-	local map=${3}
+	local version=${KV_FULL}
+	local image=${KV_DIR}/$(dist-kernel_get_image_path)
+	local map=${KV_DIR}/System.map
+	local dir=${EROOT}/boot
+
+	local arg counter=0 extrargs=( --verbose )
+	for arg in "${@}"; do
+		case "${arg}" in
+			-*) extrargs+=( "${arg}" );;
+			*)
+				let counter=counter+1
+				if [[ ${counter} -eq 1 ]]; then
+					version=${arg}
+				elif [[ ${counter} -eq 2 ]]; then
+					image=${arg}
+				elif [[ ${counter} -eq 3 ]]; then
+					map=${arg}
+				elif [[ ${counter} -eq 4 ]]; then
+					dir=${arg}
+				else
+					die "${FUNCNAME}: Too many arguments"
+				fi
+			;;
+		esac
+	done
+
+	local installkernel_args=(
+		"${version}" "${image}" "${map}" "${dir}"
+	)
+	# Versions prior to 56 have no support for parsing optional args.
+	if has_version ">=sys-kernel/installkernel-56"; then
```

Comment:
Is it actually safe to conditionally ignore additional flags?

Reply:
Yes, because so far the only optional arguments supported are `--help`, `--verbose` (which used to be the default anyway), `--(no-)systemd`(to override the use of `kernel-install`), and `--all`. We could add more in the future, but for now `--all` is really the only one of any relevance to the eclass/ebuilds. For older versions of installkernel the logic remains unchanged, the specified kernel version is reinstalled and `--all` is ignored.